src/Transit/Internal/Pipeline.hs

```diff
@@ -0,0 +1,132 @@
+                   FilePath
+                -> Int
+                -> TCPEndpoint
+                -> SecretBox.Key
+                -> C.ConduitM a c m (Text, ())
+receivePipeline fp len (TCPEndpoint s) key =
+    CN.sourceSocket s
+    .| assembleRecordC
+    .| decryptC key
+    .| CB.isolate len
+    .| sha256PassThroughC `C.fuseBoth` C.sinkFileCautious fp
+
+encryptC :: Monad m => SecretBox.Key -> C.ConduitT ByteString ByteString m ()
+encryptC key = go Saltine.zero
+  where
+    go nonce = do
+      b <- C.await
+      case b of
+        Nothing -> return ()
+        Just chunk -> do
+          let cipherText = encrypt key nonce chunk
+              cipherTextSize = toLazyByteString (word32BE (fromIntegral (BS.length cipherText)))
+          C.yield (toS cipherTextSize)
+          C.yield cipherText
+          go (Saltine.nudge nonce)
+
+decryptC :: MonadIO m => SecretBox.Key -> C.ConduitT ByteString ByteString m ()
+decryptC key = loop
+  where
+    loop = do
+      b <- C.await
+      case b of
+        Nothing -> return ()
+        Just bs -> do
+          let (nonceBytes, ciphertext) = BS.splitAt boxNonce bs
+              nonce = fromMaybe (panic "unable to decode nonce") $
+                Saltine.decode nonceBytes
+              maybePlainText = SecretBox.secretboxOpen key nonce ciphertext
+          case maybePlainText of
+            Just plaintext -> do
+              C.yield plaintext
+              loop
+            Nothing -> throwIO (CouldNotDecrypt "SecretBox failed to open")
+
+sha256PassThroughC :: (Monad m) => C.ConduitT ByteString ByteString m Text
+sha256PassThroughC = go $! Hash.hashInitWith SHA256
+  where
+    go :: (Monad m) => Hash.Context SHA256 -> C.ConduitT ByteString ByteString m Text
+    go ctx = do
+      b <- C.await
+      case b of
+        Nothing -> return $! show (Hash.hashFinalize ctx)
+        Just bs -> do
+          C.yield bs
+          go $! Hash.hashUpdate ctx bs
+
+-- | The decryption conduit computation would succeed only if a complete
+-- bytestream that represents an encrypted block of data is given to it.
+-- However, the upstream elements may chunk the data for which one may not
+-- have control of. The encrypted packet on the wire has a 4-byte length
+-- header, so we could first read it and assemble a complete encrypted
+-- block into downstream.
+assembleRecordC :: Monad m => C.ConduitT ByteString ByteString m ()
+assembleRecordC = do
+  hdr <- getChunk 4
+  let len = runGet getWord32be (BL.fromStrict hdr)
+  packet <- getChunk (fromIntegral len)
+  C.yield packet
+  assembleRecordC
+  where
+    getChunk :: Monad m => Int -> C.ConduitT ByteString ByteString m ByteString
+    getChunk size = go size BB.empty
+    go :: Monad m => Int -> BB.Builder -> C.ConduitT ByteString ByteString m ByteString
+    go size res = do
+      let residue = BL.toStrict . BB.toLazyByteString $ res
+      b <- C.await
+      case b of
+        Nothing -> return residue
```

Comment:
What happens if `residue` is shorter than `size`?  Would it be better for `getChunk` to signal an error in this case?  Or is the short return signal enough?

My naive suspicion is that encoding the exception case some way other than a short return would be better - some way that allows application code to leverage the type system to handle these cases instead of requiring that it remembers that it needs to check for a short result if that needs special handling (which I imagine it often would - and it does in this particular case).

Reply:
Good point. Thanks. I will address this in a followup commit after merging.